_posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown

```diff
@@ -0,0 +1,192 @@
+
+## 2. Gems
+
+You can add the aptly named [rails4_upgrade gem](https://github.com/alindeman/rails4_upgrade)
+to your Rails 3 project's Gemfile and find gems which you'll need to update:
+
+```
+➜  myproject git:(develop) ✗ bundle exec rake rails4:check
+
+** GEM COMPATIBILITY CHECK **
++------------------------------------------+----------------------------+
+| Dependency Path                          | Rails Requirement          |
++------------------------------------------+----------------------------+
+| devise 2.1.4                             | railties ~> 3.1            |
+| devise-encryptable 0.2.0 -> devise 2.1.4 | railties ~> 3.1            |
+| friendly_id 4.0.10.1                     | activerecord < 4.0, >= 3.0 |
+| strong_parameters 0.2.3                  | actionpack ~> 3.0          |
+| strong_parameters 0.2.3                  | activemodel ~> 3.0         |
+| strong_parameters 0.2.3                  | activesupport ~> 3.0       |
+| strong_parameters 0.2.3                  | railties ~> 3.0            |
++------------------------------------------+----------------------------+
+```
+
+Instead of going through your currently bundled gems or `Gemfile.lock` manually,
+you get a report of the gems you need to upgrade.
+
+## 3. Config
+
+Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task),
+which you can use as a guideline as explained thoroughly in
+[this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html)
+to get rid of unnecessary code or monkey-patches in your config files and
+initializers, specially if your Rails 3 app was running on Rails 2.
+
+As an alternative, check out [RailsDiff](http://railsdiff.org/3.2.22.5/4.0.13),
+which provides an overview of the changes in a basic Rails app between 3.2 and
+4.0 (or any other source/target versions).
+
+If you're feeling adventurous, you
+can give [this script](https://github.com/bsodmike/upgrade_rails_3.2.12_to_4.0.0.beta1)
+a try. It attempts to apply [this git patch](https://github.com/bsodmike/upgrade_rails_3.2.12_to_4.0.0.beta1/blob/master/upgrade/upgrade.patch)
+(similar to the patch shown on RailsDiff) to your Rails app to migrate from 3.2
+to 4.0. However, I don't recommend this for complex or mature apps, as there
+will be plenty of conflicts.
+
+## 4. Application code
+
+### a. Models
+
+- All dynamic finder methods except for `.find_by_...` are deprecated:
+
+```ruby
+# before:
+Authentication.find_all_by_provider_and_uid(provider, uid)
+
+# after:
+Authentication.where(provider: provider, uid: uid)
+```
+
+You can regain usage of these finders by adding the gem
+[activerecord-deprecated_finders](https://github.com/rails/activerecord-deprecated_finders)
+
+- ActiveRecord scopes now need a lambda:
+
+```ruby
+# before:
+default_scope where(deleted_at: nil)
+
+# after:
+default_scope { where(deleted_at: nil) }
```

Comment:
@mauro-oto Maybe you can post a link to your previous article? 😊  https://www.ombulabs.com/blog/ruby/rails/best-practices/why-using-default-scope-is-a-bad-idea.html